### ISSUE TITLE
StrictBool, StrictStr etc types are not drop in replacements for normal types

### ISSUE THREAD

**Issue by spookylukey** [pydantic/pydantic#2329]
Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.7.3
            pydantic compiled: True
                 install path: /home/luke/.virtualenvs/rpr/lib/python3.9/site-packages/pydantic
               python version: 3.9.1 (default, Jan 26 2021, 09:48:27)  [GCC 9.3.0]
                     platform: Linux-5.4.0-58-generic-x86_64-with-glibc2.31
     optional deps. installed: []
```

In places where you are trying to tighten validation by going from `int` to `StrictInt` etc, these types fail as drop in replacements in several ways. Specifically, if you call the types directly, they produce instances of `StrictInt` etc, instead of instances of `int` etc. This produces some incompatibilities.

For `StrictBool`, `repr()` output:

```py
>>> from pydantic import StrictBool
>>> repr(StrictBool(False))
'0'
# Compared with:
>>> repr(bool(False))
'False'
```

For `StrictBool`, `isinstance()` checks:
```py
>>> from pydantic import StrictBool
>>> isinstance(StrictBool(False), bool)
False
```

For `StrictBool`, any code that does `is` checks, which is a legitimate technique, especially when you want to check for `False` as distinct from `None`:

```py
>>> StrictBool(False) is False
False
# Compared with
>>> bool(False) is False
True
```

For all of `StrictInt`, `StrictStr`, `StrictBool`, any code that does `type()` checking will behave differently. This is arguably a bug of that code (it should be doing `isinstance()`), but it is worth asking whether the current behaviour is helpful. When you change the type to `StrictInt` instead of `int`, you do so only for different validation behaviour.

## Use case - type driven deserialization

Here is a fuller use case - doing type driven deserialization, similar to `BaseModel.parse_obj`, but with my own logic. 
Starting with these classes:

```py
from enum import Enum
from pydantic import BaseModel


class Color(Enum):
    RED = 'r'
    BLUE = 'b'


class Flag(BaseModel):
    color: Color
    square: StrictBool

```
Suppose we have a serialized instance of `Flag` as JSON string `{"color": "r", "square": true}` (just like `.json()`), which becomes the Python dict `{"color": "r", "square": True}` after JSON decoding. So we make a `deserializer()` like this:

```py
def deserialize(data, cls):
    return cls(**{name: member_cls(data[name]) for name, member_cls in cls.__annotations__.items()})
```

(This is obviously limited, but with some small enhancements we can handle lists etc and complex nested types)

When we try to run it, however, we get this:
```py
>>> deserialize({'color': 'r', 'square': True}, cls=Flag)
ValidationError: 1 validation error for Flag
square
  value is not a valid boolean (type=value_error.strictbool)
```

Here, we are falling foul of Pydantic's validation - the `cls()` call in `deserialize()` returned a `StrictBool` instance, which fails an `isinstance(x, bool)` check, as noted above.

BTW, this is pretty difficult to debug since at a Python prompt, the objects `repr()` makes them look like the simple builtin types, hiding what is actually happening, which is another disadvantage of the current behaviour.

## Use case - automatic test data generation with hypothesis

```py
>>> from hypothesis.strategies import from_type
>>> from_type(Flag).example()
ValidationError: 1 validation error for Flag
square
  value is not a valid boolean (type=value_error.strictbool)
```
This is essentially the same issue as above.

By contrast, `from_type` works great if we go from `StictBool` to `bool`.

(We don't have a problem with `StrictInt` here, but I do notice that hypothesis always generates `0` for `StrictInt`, but not for `int`. This might be worth addressing at the same time if possible).

## Proposal

The `__new__` methods of these types should be overridden to call the base type constructor. I think this would solve all these issues.

**Comment by PrettyWood** [comment]
Hi @spookylukey and thanks a lot for reporting.
I agree with both the issue and the proposal: adding a basic `__new__` method is probably the way to go.
PR welcome :)

**Comment by Zac-HD** [comment]
> (We don't have a problem with `StrictInt` here, but I do notice that hypothesis always generates `0` for `StrictInt`, but not for `int`. This might be worth addressing at the same time if possible).

This indicates that we're generating `StrictInt` literally by calling the construction (`StrictInt()`, via `st.builds(StrictInt)`), which is not the intended behaviour...  `StrictBool` is handled [here](https://github.com/samuelcolvin/pydantic/blob/619ff261c9bdc2a612ef2fcfa17a005dfefbcf21/pydantic/_hypothesis_plugin.py#L165), but `StrictInt` *should* be handled by [`ConstrainedNumberMeta`](https://github.com/samuelcolvin/pydantic/blob/619ff261c9bdc2a612ef2fcfa17a005dfefbcf21/pydantic/types.py#L139-L148) calling `_register()`.

If it's not, something #2473-related might be happening?

**Comment by xtoeffel** [comment]
Related to this:
```python
import pydantic as pyd
class StrictBoolModel(pyd.BaseModel):
    strict_bool: pyd.StrictBool
StrictBoolModel(strict_bool=pyd.StrictBool(True))
```
```
pydantic.error_wrappers.ValidationError: 1 validation error for StrictBoolModel
strict_bool
  value is not a valid boolean (type=value_error.strictbool)
```
This is unexpected because passing a `StrictBool` to create a `StrictBool` is valid:
```python
pyd.StrictBool(pyd.StrictBool(True))
# out: 1
```

**Comment by samuelcolvin** [comment]
Fixed by #4516, will be fixed in V2.
